agent/agentexec/exec.go

```diff
@@ -20,60 +20,101 @@ const (
-// the provided command if CODER_PROC_PRIO_MGMT is set, otherwise a normal exec.Cmd
-// is returned. All instances of exec.Cmd should flow through this function to ensure
-// proper resource constraints are applied to the child process.
-func CommandContext(ctx context.Context, cmd string, args ...string) (*exec.Cmd, error) {
-	cmd, args, err := agentExecCmd(cmd, args...)
-	if err != nil {
-		return nil, xerrors.Errorf("agent exec cmd: %w", err)
-	}
-	return exec.CommandContext(ctx, cmd, args...), nil
-}
+	// unset is set to an invalid value for nice and oom scores.
+	unset = -2000
+)
 
-// PTYCommandContext returns an pty.Cmd that calls "coder agent-exec" prior to exec'ing
-// the provided command if CODER_PROC_PRIO_MGMT is set, otherwise a normal pty.Cmd
-// is returned. All instances of pty.Cmd should flow through this function to ensure
-// proper resource constraints are applied to the child process.
-func PTYCommandContext(ctx context.Context, cmd string, args ...string) (*pty.Cmd, error) {
-	cmd, args, err := agentExecCmd(cmd, args...)
-	if err != nil {
-		return nil, xerrors.Errorf("agent exec cmd: %w", err)
-	}
-	return pty.CommandContext(ctx, cmd, args...), nil
+var DefaultExecer Execer = execer{}
+
+// Execer defines an abstraction for creating exec.Cmd variants. It's unfortunately
+// necessary because we need to be able to wrap child processes with "coder agent-exec"
+// for templates that expect the agent to manage process priority.
+type Execer interface {
+	// CommandContext returns an exec.Cmd that calls "coder agent-exec" prior to exec'ing
+	// the provided command if CODER_PROC_PRIO_MGMT is set, otherwise a normal exec.Cmd
+	// is returned. All instances of exec.Cmd should flow through this function to ensure
+	// proper resource constraints are applied to the child process.
+	CommandContext(ctx context.Context, cmd string, args ...string) *exec.Cmd
+	// PTYCommandContext returns an pty.Cmd that calls "coder agent-exec" prior to exec'ing
+	// the provided command if CODER_PROC_PRIO_MGMT is set, otherwise a normal pty.Cmd
+	// is returned. All instances of pty.Cmd should flow through this function to ensure
+	// proper resource constraints are applied to the child process.
+	PTYCommandContext(ctx context.Context, cmd string, args ...string) *pty.Cmd
 }
 
-func agentExecCmd(cmd string, args ...string) (string, []string, error) {
+func NewExecer() (Execer, error) {
 	_, enabled := os.LookupEnv(EnvProcPrioMgmt)
 	if runtime.GOOS != "linux" || !enabled {
-		return cmd, args, nil
+		return DefaultExecer, nil
 	}
 
 	executable, err := os.Executable()
 	if err != nil {
-		return "", nil, xerrors.Errorf("get executable: %w", err)
+		return nil, xerrors.Errorf("get executable: %w", err)
 	}
 
 	bin, err := filepath.EvalSymlinks(executable)
 	if err != nil {
-		return "", nil, xerrors.Errorf("eval symlinks: %w", err)
+		return nil, xerrors.Errorf("eval symlinks: %w", err)
+	}
+
+	oomScore, ok := envValInt(EnvProcOOMScore)
```

Comment:
I see there's no validation, but what happens if this is set to <-20 or >20, will everything fail to exec, will it silently be ignored, or something else?

Also not sure if system level prio max/min can be adjusted, typically it's 20 anyway. 🤔

Reply:
It'll fail to adjust the priority which will print a message to stderr but it won't fatally kill the command